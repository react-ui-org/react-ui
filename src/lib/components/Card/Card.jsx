import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Card.scss';

export const Card = ({
  children,
  dense,
  disabled,
  id,
  raised,
  color,
}) => (
  <div
    className={classNames(
      styles.root,
      getRootColorClassName(color, styles),
      dense && styles.rootDense,
      raised && styles.rootRaised,
      disabled && styles.isDisabled,
    )}
    id={id}
  >
    {children}
  </div>
);

Card.defaultProps = {
  color: 'light',
  dense: false,
  disabled: false,
  id: undefined,
  raised: false,
};

Card.propTypes = {
  /**
   * Slot for individual card elements that build up the inner layout:
   * * `CardBody`
   * * `CardFooter`
   * * `ScrollView`
   */
  children: PropTypes.node.isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the card.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Make the card more compact.
   */
  dense: PropTypes.bool,
  /**
   * If `true`, the card will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Add shadow to pull the card above surface.
   */
  raised: PropTypes.bool,
};

export const CardWithContext = withProviderContext(Card, 'Card');

export default CardWithContext;
