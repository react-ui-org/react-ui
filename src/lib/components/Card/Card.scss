// 1. Retain equal card widths in flex and grid layouts independently on their content.

@use "../../styles/theme/colors";
@use "theme";
@use "tools";

.root {
    --rui-local-padding: #{theme.$padding};

    display: flex;
    flex-direction: column;
    min-width: 0; // 1.
    color: var(--rui-local-color);
    border: theme.$border-width solid var(--rui-local-border-color);
    border-radius: theme.$border-radius;
    background-color: var(--rui-local-background-color);
}

.body {
    flex: 1 0 auto;
    padding: var(--rui-local-padding);
}

.footer {
    padding: var(--rui-local-padding);
}

.rootDense {
    --rui-local-padding: #{theme.$dense-padding};
}

.rootRaised {
    box-shadow: theme.$raised-box-shadow;
}

.rootColorPrimary {
    @include tools.color(primary);
}

.rootColorSecondary {
    @include tools.color(secondary);
}

.rootColorSuccess {
    @include tools.color(success);
}

.rootColorWarning {
    @include tools.color(warning);
}

.rootColorDanger {
    @include tools.color(danger);
}

.rootColorHelp {
    @include tools.color(help);
}

.rootColorInfo {
    @include tools.color(info);
}

.rootColorNote {
    @include tools.color(note);
}

.rootColorLight {
    @include tools.color(light);
}

.rootColorDark {
    @include tools.color(dark);
}

.isDisabled {
    background-color: theme.$disabled-background-color;
    opacity: theme.$disabled-opacity;
}
