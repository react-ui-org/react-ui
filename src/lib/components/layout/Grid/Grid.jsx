import flattenChildren from 'react-keyed-flatten-children';
import PropTypes from 'prop-types';
import React from 'react';
import styles from './Grid.scss';

export const Grid = (props) => {
  const {
    children,
    id,
    ...other
  } = props;

  if (!props.children) {
    return null;
  }

  return (
    <div
      id={id}
      className={styles.root}
      {...other}
    >
      {flattenChildren(children).map((child) => {
        if (!React.isValidElement(child)) {
          return null;
        }

        return React.cloneElement(child);
      })}
    </div>
  );
};

Grid.defaultProps = {
  children: null,
  id: undefined,
};

Grid.propTypes = {
  /**
   * Items to be aligned in the grid.
   */
  children: PropTypes.node,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
};

export default Grid;
