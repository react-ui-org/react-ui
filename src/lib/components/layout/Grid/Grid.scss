@import '../../../styles/settings/layouts';
@import '../../../styles/tools/breakpoints';
@import './theme';

/* autoprefixer grid: off */
.root {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto;
  grid-gap: $grid-gap;
  margin-bottom: $layout-common-bottom-spacing;

  @include breakpoint-up(sm) {
    grid-template-columns: repeat(auto-fit, minmax($grid-item-min-width, $grid-item-max-width));
    justify-content: start;
  }
}
