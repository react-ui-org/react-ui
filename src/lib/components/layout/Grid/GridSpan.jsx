import PropTypes from 'prop-types';
import React from 'react';
import { generateResponsiveCustomProperties } from './helpers/generateResponsiveCustomProperties';
import styles from './Grid.scss';

export const GridSpan = ({
  children,
  columns,
  id,
  rows,
  ...other
}) => {
  if (!children) {
    return null;
  }

  return (
    <div
      id={id}
      className={styles.span}
      style={{
        ...generateResponsiveCustomProperties(columns, 'column-span'),
        ...generateResponsiveCustomProperties(rows, 'row-span'),
      }}
      {...other}
    >
      {children}
    </div>
  );
};

/* Breakpoints are easier to work with when ordered according to their value, not name. */
/* eslint-disable sort-keys */

GridSpan.defaultProps = {
  children: null,
  columns: undefined,
  id: undefined,
  rows: undefined,
};

GridSpan.propTypes = {
  /**
   * Items to be aligned in the grid.
   */
  children: PropTypes.node,
  /**
   * Number of columns to span.
   */
  columns: PropTypes.oneOf([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Number of rows to span.
   */
  rows: PropTypes.oneOf([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
};

export default GridSpan;
