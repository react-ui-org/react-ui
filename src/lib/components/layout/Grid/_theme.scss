$grid-item-min-width: var(--rui-grid-item-min-width);
$grid-item-max-width: var(--rui-grid-item-max-width);
$grid-gap: var(--rui-grid-gap);
