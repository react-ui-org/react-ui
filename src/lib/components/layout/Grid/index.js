export { default as Grid } from './Grid';
export { default as GridSpan } from './GridSpan';
