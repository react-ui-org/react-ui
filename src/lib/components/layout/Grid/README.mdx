---
name: Grid
menu: 'Components: Layouts'
route: /components/layout/grid
---

# Grid

import {
  Playground,
  Props,
} from 'docz'
import { Placeholder } from '../../../../docs/_components/Placeholder/Placeholder'
import { Grid } from './Grid'

The responsive Grid layout aligns content into an organized grid.

## Basic Usage

To implement the Grid component, you need to import it first:

```js
import { Grid } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Grid>
    <Placeholder bordered>Grid item</Placeholder>
    <Placeholder bordered>Grid item</Placeholder>
    <Placeholder bordered>Grid item</Placeholder>
    <Placeholder bordered>Grid item</Placeholder>
  </Grid>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- To align your items in the grid, **simply wrap** them with the Grid
  component. Unlike other grid frameworks, **no additional markup** like
  GridItem or Cell is necessary. This is possible thanks to the
  [CSS grid layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout),
  the right tool for two-dimensional layouts.

- If you need your items to have **equal height** even with content of varying
  length, it may be necessary to set `height: 100%` on them.

ðŸ‘‰ The default column width algorithm is `repeat(auto-fit, minmax(300px, auto))`
and it can be [customized](/customize/theming) with CSS custom properties.
More layout options are [coming soon](https://github.com/react-ui-org/react-ui/issues/149)!

## API

<Props of={Grid} />

---

Next: [List â†’](/components/layout/list)
