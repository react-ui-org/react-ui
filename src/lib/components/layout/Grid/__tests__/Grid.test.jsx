import React from 'react';
import { shallow } from 'enzyme';
import { Card } from '../../../ui/Card/Card';
import { Grid } from '../Grid';

describe('rendering', () => {
  it('renders correctly with no children', () => {
    const tree = shallow((
      <Grid />
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a single child', () => {
    const tree = shallow((
      <Grid>
        <Card>content</Card>
      </Grid>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <Grid>
        <Card>content 1</Card>
        <Card>content 2</Card>
        <Card>content 3</Card>
      </Grid>
    ));

    expect(tree).toMatchSnapshot();
  });
});
