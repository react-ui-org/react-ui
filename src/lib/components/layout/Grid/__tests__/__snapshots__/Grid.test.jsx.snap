// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root"
>
  <Card
    dense={false}
    disabled={false}
    key=".0"
    raised={false}
    type={null}
    variant="flat"
  >
    content
  </Card>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root"
>
  <Card
    dense={false}
    disabled={false}
    key=".0"
    raised={false}
    type={null}
    variant="flat"
  >
    content 1
  </Card>
  <Card
    dense={false}
    disabled={false}
    key=".1"
    raised={false}
    type={null}
    variant="flat"
  >
    content 2
  </Card>
  <Card
    dense={false}
    disabled={false}
    key=".2"
    raised={false}
    type={null}
    variant="flat"
  >
    content 3
  </Card>
</div>
`;

exports[`rendering renders correctly with no children 1`] = `""`;
